pkg/destroy/azure/azure.go

```diff
@@ -217,3 +235,68 @@ func deleteResourceGroup(ctx context.Context, client resources.GroupsGroupClient
 func wasNotFound(resp *http.Response) bool {
 	return resp != nil && resp.StatusCode == http.StatusNotFound
 }
+
+func deleteApplicationRegistrations(ctx context.Context, appClient graphrbac.ApplicationsClient, spClient graphrbac.ServicePrincipalsClient, logger logrus.FieldLogger, infraID string) error {
+	errorList := []error{}
+
+	tag := fmt.Sprintf("kubernetes.io_cluster.%s=owned", infraID)
+	servicePrincipals, err := getServicePrincipalsByTag(ctx, spClient, tag, infraID)
+	if err != nil {
+		return errors.Wrap(err, "failed to gather list of Service Principals by tag")
+	}
+
+	for _, sp := range servicePrincipals {
+		logger = logger.WithField("appID", *sp.AppID)
+		appFilter := fmt.Sprintf("appId eq '%s'", *sp.AppID)
+		appResults, err := appClient.List(ctx, appFilter)
+		if err != nil {
+			errorList = append(errorList, err)
+			continue
+		}
+
+		apps := appResults.Values()
+		if len(apps) != 1 {
+			msg := fmt.Sprintf("should have recieved only a single result matching AppID, received %d instead", len(apps))
+			errorList = append(errorList, errors.New(msg))
+			continue
+		}
+
+		_, err = appClient.Delete(ctx, *apps[0].ObjectID)
+		if err != nil {
+			errorList = append(errorList, err)
+			continue
+		}
+		logger.Info("deleted")
+	}
+
+	return utilerrors.NewAggregate(errorList)
+}
+
+func getServicePrincipalsByTag(ctx context.Context, spClient graphrbac.ServicePrincipalsClient, matchTag, infraID string) ([]graphrbac.ServicePrincipal, error) {
+	matchedSPs := []graphrbac.ServicePrincipal{}
+
+	infraFilter := fmt.Sprintf("startswith(displayName,'%s')", infraID)
+	spResults, err := spClient.List(ctx, infraFilter)
+	if err != nil {
+		return matchedSPs, err
+	}
+
+	for {
```

Comment:
use somethings like https://github.com/openshift/installer/blob/045402165b7aa5cb95ab050d4ca0827016e05dd1/pkg/asset/installconfig/azure/dns.go#L121-L132

Reply:
:+1: i much prefer this. will change it